lib/async_hooks.js

```diff
@@ -98,24 +103,19 @@ class AsyncHook {
     if (hooks_array.includes(this))
       return;
 
-    if (!setupHooksCalled) {
-      setupHooksCalled = true;
-      // Setup the callbacks that node::AsyncWrap will call when there are
-      // hooks to process. They use the same functions as the JS embedder API.
-      async_wrap.setupHooks({ init,
-                              before: emitBeforeN,
-                              after: emitAfterN,
-                              destroy: emitDestroyN });
-    }
-
     // createHook() has already enforced that the callbacks are all functions,
     // so here simply increment the count of whether each callbacks exists or
     // not.
     hook_fields[kInit] += +!!this[init_symbol];
     hook_fields[kBefore] += +!!this[before_symbol];
     hook_fields[kAfter] += +!!this[after_symbol];
     hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hook_fields[kTotals]++;
```

Comment:
I don't understand what @matthewloring mean with: 

> If listeners are added later then it might be good to have PromiseWraps available to resolve parent promise ids.

Is this related to https://github.com/nodejs/node/issues/13427?